### ISSUE TITLE
Got Permission Denial: get/set setting for user asks to run as user -2 but is calling from user 0

### ISSUE THREAD

**Issue by attawit** [gotev/android-upload-service#283]
I have used ftp-upload-service, Basically it have been worked well for few days, But there have some error as followed by not changed of the code. Even after I have used simple ftp upload code: 

`07-03 11:28:30.802 30377-30377/com.jinn.mutita.photoImx E/UPLOAD START File: /data/user/0/com.jinn.mutita.photoImx/app_imageDir/20170316_190501.jpg
07-03 11:28:31.187 4470-14525/? E/DatabaseUtils: Writing exception to parcel
                                                 java.lang.SecurityException: Permission Denial: get/set setting for user asks to run as user -2 but is calling from user 0; this requires android.permission.INTERACT_ACROSS_USERS_FULL
                                                     at com.android.server.am.ActivityManagerService.handleIncomingUser(ActivityManagerService.java:22048)
                                                     at android.app.ActivityManager.handleIncomingUser(ActivityManager.java:3374)
                                                     at com.android.providers.settings.SettingsProvider.resolveCallingUserIdEnforcingPermissionsLocked(SettingsProvider.java:1289)
                                                     at com.android.providers.settings.SettingsProvider.getSystemSetting(SettingsProvider.java:949)
                                                     at com.android.providers.settings.SettingsProvider.call(SettingsProvider.java:301)
                                                     at android.content.ContentProvider$Transport.call(ContentProvider.java:411)
                                                     at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:283)
                                                     at android.os.Binder.execTransact(Binder.java:453)
07-03 11:28:31.187 4470-14524/? E/DatabaseUtils: Writing exception to parcel
                                                 java.lang.SecurityException: Permission Denial: get/set setting for user asks to run as user -2 but is calling from user 0; this requires android.permission.INTERACT_ACROSS_USERS_FULL
                                                     at com.android.server.am.ActivityManagerService.handleIncomingUser(ActivityManagerService.java:22048)
                                                     at android.app.ActivityManager.handleIncomingUser(ActivityManager.java:3374)
                                                     at com.android.providers.settings.SettingsProvider.resolveCallingUserIdEnforcingPermissionsLocked(SettingsProvider.java:1289)
                                                     at com.android.providers.settings.SettingsProvider.getSystemSetting(SettingsProvider.java:949)
                                                     at com.android.providers.settings.SettingsProvider.call(SettingsProvider.java:301)
                                                     at android.content.ContentProvider$Transport.call(ContentProvider.java:411)
                                                     at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:283)
                                                     at android.os.Binder.execTransact(Binder.java:453)
07-03 11:28:31.192 4470-14470/? E/DatabaseUtils: Writing exception to parcel
                                                 java.lang.SecurityException: Permission Denial: get/set setting for user asks to run as user -2 but is calling from user 0; this requires android.permission.INTERACT_ACROSS_USERS_FULL
                                                     at com.android.server.am.ActivityManagerService.handleIncomingUser(ActivityManagerService.java:22048)
                                                     at android.app.ActivityManager.handleIncomingUser(ActivityManager.java:3374)
                                                     at com.android.providers.settings.SettingsProvider.resolveCallingUserIdEnforcingPermissionsLocked(SettingsProvider.java:1289)
                                                     at com.android.providers.settings.SettingsProvider.getSystemSetting(SettingsProvider.java:949)
                                                     at com.android.providers.settings.SettingsProvider.call(SettingsProvider.java:301)
                                                     at android.content.ContentProvider$Transport.call(ContentProvider.java:411)
                                                     at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:283)
                                                     at android.os.Binder.execTransact(Binder.java:453)
07-03 11:28:31.297 2975-4297/? E/MediaPlayerService: WFD client is not created 
07-03 11:28:31.332 2975-30591/? E/OMXNodeInstance: setConfig(ae:google.vorbis.decoder, ConfigPriority(0x6f800002)) ERROR: Undefined(0x80001001)
07-03 11:28:36.212 2961-4281/? E/NetlinkEvent: NetlinkEvent::FindParam(): Parameter 'TIME_NS' not found
07-03 11:28:36.212 2961-4281/? E/NetlinkEvent: NetlinkEvent::FindParam(): Parameter 'UID' not found
07-03 11:28:37.497 4470-13935/? E/Watchdog: !@Sync 339 [07-03 11:28:37.502]
`
### Version info
* Android Upload Service version: 3.2.4
* Android version and API version: 23

```java
// the code you wrote in your Application subclass here
```

### Request code:
`                 ImageView vs = (ImageView) findViewById(R.id.photo_temp);
                 Bitmap bitm = BitmapFactory.decodeFile(mFile.get(0).getPath());

                 vs.setImageBitmap(bitm);
                 Log.e("UPLOAD START File",mFile.get(0).getPath());
                 FTPUploadRequest uploadId =
                        new FTPUploadRequest(PhotoImxMainActivity.this, Template.FTP.SERVER, Template.FTP.PORT)
                                .setUsernameAndPassword(Template.FTP.USER, Template.FTP.PASSWORD)
                                .addFileToUpload(mFile.get(0).getPath(), Template.FTP.UPLOAD_DIR)
                                .setNotificationConfig(new UploadNotificationConfig())
                                .setCreatedDirectoriesPermissions(new UnixPermissions("777"))
                                .setSocketTimeout(5000)
                                .setConnectTimeout(5000);
                 uploadId.startUpload() ;`
```

However, if I have remove .setNotificationConfig, There are no Exception Error as above but It still can not work also. Thanks you for you help

**Comment by gotev** [comment]
The error seems unrelated to the library. I see securuty exceptions from other components of your app. Check if you have correctly requested runtime permissions on Android 6.0. I suggest you to try the demo app with your server and try the same things, so you can spot where tge real problem is

**Comment by attawit** [comment]
I have checked on the requested runtime permissions already. And FYI, if I remove    .setNotificationConfig(new UploadNotificationConfig()) from the Request, this Writing exception to parcel java.lang.SecurityException are gone.   ...

Let be try Demo App but frankly speaking, I have issue on set up Butterknift Bind on the demo code which found some error on my Android Studio.  Let me try again...

**Comment by attawit** [comment]
Still have the same error Security Exception even on the demo code.

**Comment by gotev** [comment]
Ok, so the problem is in the way you're getting the URI to the file. You got to have a content uri, which has no problems on Android 6.0+ with Upload Service. The issue is because of stricter policies applied by the Android platforms that you have to follow. [This](https://github.com/gotev/android-upload-service/issues/269#issuecomment-304250700) can help you

**Comment by attawit** [comment]
I have tried this.  And by the way the file have been new created by compressing from the original uri.  content and the file was written in the sandbox of the app which I have test before that It is not empty or unreachable .  Let be try to understand Async with parcel it may related issue.

**Comment by gotev** [comment]
It would be great if you could make a fork and replicate your problem in
the upload service demo app, so I could help you and then I could use it as
a test case against new versions.

Il 05 lug 2017 6:08 PM, "attawit" <notifications@github.com> ha scritto:

**Comment by attawit** [comment]
I have tried to test when I have changed to API 22, during that time I found some mistake on server IP "192.122.123.22 " there is a space after 22, After I have removed that But I have removed the .setNotificationConfig(new UploadNotificationConfig()) also. It seems working. Then I'd like to regenerate the error again. By changing back to API 23, I put back everything but can not generate that error. But that Exception still occurs related to facebook.  Thank you , I will buy a beer if you are in thailand :)

**Comment by gotev** [comment]
A String trimmer in the request configuration will be helpful. I'll add it in the next release. I suggest you to re-enable the notification config, otherwise you could not benefit from the foreground execution of the upload task, resulting in lower upload rate and higher probability that Android kills the task if it needs memory or to save battery. If you look at the upload service source, I've documented everything in detail. Basically android wants a persistent running notification to enable foreground execution for the service.

**Comment by attawit** [comment]
You are correct, this have been related to the case of Android kills the task while it needs memory also.

**Comment by attawit** [comment]
BTW, do you ever seen problem with writing file in root directory instead of directory when put the parameter in function. Is this related to the ftp daemon which may not be supported or not?

**Comment by gotev** [comment]
No problems at all, you can try the demo ftp server provided with the FTP module. Where tou can write on the server depends entirely by the policy set for the user in the FTP configuration
